### DISCUSSION TITLE
Allow 'return' keyword in void(...) functions to accept any valid expression as a parameter

### DISCUSSION THREAD

**Discussion by apalamart** [dotnet/csharplang#5558] [Language Ideas]
Hi,
The suggestion is to discuss a small change that can allow a 'return' keyword in a void(...) function to accept any valid expression or code block as a parameter and execute that expression before returning execution to a caller.

The following blocks of code can explain what is proposed:
```cs
void SomeFunc(object param)
{
  if(!(param is string tmpStr))
  {
    MessageBox.Show("Param should be a string");
    return;
  }

  if(tmpStr != "action")
  {
    Logger.Log("A wrong action specified");
    MessageBox.Show("A wrong action specified");
    return;
  }

  //... do something
}
```
at the same time, the same behavior could be expressed as follows:
```cs
void SomeFunc(object param)
{
  if(!(param is string tmpStr))
    return MessageBox.Show("Param should be a string");

  if(tmpStr != "action")
    return {
      Logger.Log("A wrong action specified");
      MessageBox.Show("A wrong action specified");
    }

  //... do something
}
```
However, the latter is not a valid code and is described in [CS0127](https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs0127?f1url=%3FappId%3Droslyn%26k%3Dk(CS0127)).

UPD:

The example above can be confusing because of calling MessageBox-es on return, which is subject to refactoring,
So let's say that we need to call other method before return;
Another good comment is about understanding the code, so that it does not confuse a reviewer to be returning the result or the expression to a caller.
So the return expression could look like following:

```cs
if(some_validation)
  return (void)ProcessTheDataOtherWay(data);

```
It would be nice to allow this kind of syntax. Hopefully, this could make the code more compact and readable.
Thank you.

**Comment by jnm2** [comment]
I would refactor these things to pull the MessageBox calls out of the logic. (That also makes the logic more susceptible to automated testing!)

```cs
Result SomeFunc(object param)
{
    if (param is not string tmpStr)
        return Result.Error("Param should be a string");

    if (tmpStr != "action")
    {
        Logger.Log("A wrong action specified");
        return Result.Error("A wrong action specified");
    }

    //... do something

    return Result.Success();
}
```

Then the caller does:

```cs
if (SomeFunc(x).IsError(out var message))
    MessageBox.Show(owner: this, caption: this.Text, message, MessageBoxButtons.OK, MessageBoxIcon.Error);
```

Here's the `Result` types as I typically define them: https://gist.github.com/jnm2/a0031a31d04d6ed4d44d8dfc9265b8cf

This is a general technique that often directly addresses the desire to use a single `return` statement for early returns rather than multiple statements inside braces.

**Reply by theunrepentantgeek** [reply]
> this could make the code more compact and readable.

Compact is seldom a goal in its own right. 

The keyword **return** is used across dozens of languages to specify the values (expressions) to return.

Your samples therefore look like returning an expression block for the caller to invoke later.

**Reply by CyrusNajmabadi** [reply]
indeed.  this would feel far too ambiguous/confusing IMO.

**Reply by apalamart** [reply]
@jnm2 I agree, may be the example I've referenced is not the best, and it's better to refactor the code the way you did.

However, the same can be applied to a method call that does not show any UI and just does a call to other void or non-void sub-routine, e.g.

```
if(some_validation)
  return ProcessTheDataOtherWay(data);

```

@theunrepentantgeek 
> Compact is seldom a goal in its own right.

So does this mean that all the syntax sugar in C# should be eliminated? 

> Your samples therefore look like returning an expression block for the caller to invoke later.

This makes sense and could be a good reason not to do this.
So will the following example make the syntax understandable?

```
if(some_validation)
  return (void)ProcessTheDataOtherWay(data);

```

Thank you guys for your answers and concern.

**Reply by CyrusNajmabadi** [reply]
> So will the following example make the syntax understandable?

IMO no.  Now i ahve to ask "what does it mean to 'return void?'

To me, there's an existing idiomatic form for this that is really easy to read and understand.  Namely your original code :)

I don't look at that and see anything that needs adjustment.

**Reply by theunrepentantgeek** [reply]
> So does this mean that all the syntax sugar in C# should be eliminated?

By no means - I did write _seldom_ instead of _never_.

Changes like expression bodied methods are syntactic sugar that's _frequently_ applicable - the overhead of of design, testing, documentation and end developer learning are more than repaid by the fact they can be pervasively applied.

Adding syntactic sugar for rare cases just means that developers will need to look up the syntax every time they encounter it, making the code cryptic, not concise.

**Reply by jnm2** [reply]
> I agree, may be the example I've referenced is not the best, and it's better to refactor the code the way you did.
>
> However, the same can be applied to a method call that does not show any UI and just does a call to other void or non-void sub-routine, e.g.

@apalamart You're right, and there are a few places here and there where I would make use of `return <void call or assignment>;`. Not many though.
